File: app/api/routers/v2.py
"""V2 API — everything the four Results screens and the ops screens read.

All numbers come from catalogued GQ queries over graph data. Commentary
endpoints RETRIEVE stored, versioned text (COMMENTARY_MODE=stored — never
generated on read; the generation workflow has its own POST).
"""
from __future__ import annotations

from fastapi import APIRouter, Query

from app.shared.responses import ok
from app.v2.drivers.service import V2DriverService
from app.v2.revenue.service import V2RevenueService

router = APIRouter(prefix="/api/v2", tags=["V2"])


# ---------------------------------------------------------------- reference

@router.get("/reference/advisors")
def advisors():
    return ok(data=V2RevenueService().advisors())


@router.get("/reference/months")
def months():
    return ok(data=V2RevenueService().months())


@router.get("/reference/product-hierarchy")
def product_hierarchy():
    return ok(data=V2RevenueService().product_hierarchy())


@router.get("/reference/driver-causes")
def driver_causes():
    return ok(data=V2RevenueService().driver_causes())


@router.get("/reference/reason-codes")
def reason_codes():
    """Eligibility reference (R1): the reason-code rows that define credited
    revenue, straight from the graph."""
    return ok(data=V2RevenueService().reason_codes())


# ---------------------------------------------------------------- trends

@router.get("/trends/revenue")
def trends_revenue(advisor_id: str, from_month: str, to_month: str):
    return ok(data=V2RevenueService().monthly_revenue(advisor_id, from_month, to_month))


@router.get("/trends/changes")
def trends_changes(advisor_id: str, from_month: str, to_month: str):
    return ok(data=V2RevenueService().revenue_changes(advisor_id, from_month, to_month))


# ---------------------------------------------------------------- insights

@router.get("/insights/chart")
def insights_chart(advisor_id: str, from_month: str, to_month: str):
    return ok(data=V2RevenueService().monthly_totals(advisor_id, from_month, to_month))


@router.get("/insights/drivers")
def insights_drivers(advisor_id: str, from_month: str, to_month: str,
                     result_limit: int = Query(default=100, le=10000)):
    return ok(data=V2DriverService().change_drivers(advisor_id, from_month, to_month, result_limit))


# ---------------------------------------------------------------- commentary (stored — never generated on read)

@router.get("/insights/commentary")
def insights_commentary(advisor_id: str, version_id: str = ""):
    from app.agents.nodes.supervisor_agent import SupervisorAgent

    return ok(data=SupervisorAgent().read_commentary(advisor_id, version_id))


@router.get("/insights/versions")
def insights_versions():
    from app.graph.client import get_graph_client
    from app.graph.queries.common import v2_served_by_tier

    result = get_graph_client().run_query("get_commentary_versions", {})
    rows = []
    for obj in result.get("results", []):
        rows = [r.get("attributes", {}) for r in obj.get("versions", [])]
    return ok(data={"versions": rows, "served_by_tier": v2_served_by_tier(result)})


@router.get("/insights/evaluations")
def insights_evaluations(version_id: str = ""):
    """LLM-as-judge verdicts (R5) for a commentary version ("" = all versions).
    Advisory only — surfaced in the UI, never a publication gate."""
    return ok(data=V2RevenueService().commentary_evaluations(version_id))


@router.post("/insights/generate")
def insights_generate(notes: str = ""):
    """Trigger the batch generation workflow — a NEW version every run; prior
    versions are never deleted. The ONLY path that reaches the LLM."""
    from app.v2.commentary.generation_workflow import run_generation

    return ok(data=run_generation(notes))


@router.get("/insights/generate/status")
def insights_generate_status():
    from app.v2.commentary.generation_workflow import get_status

    return ok(data=get_status())


@router.get("/evidence")
def evidence(driver_id: str, version_id: str = ""):
    from app.graph.client import get_graph_client
    from app.graph.queries.common import v2_served_by_tier

    result = get_graph_client().run_query(
        "get_evidence", {"driver_id": driver_id, "version_id": version_id})
    rows = []
    for obj in result.get("results", []):
        rows = [r.get("attributes", {}) for r in obj.get("evidence", [])]
    return ok(data={"evidence": rows, "served_by_tier": v2_served_by_tier(result)})


# ---------------------------------------------------------------- drill-down

@router.get("/transactions")
def transactions(advisor_id: str, month_id: str, group_id: str = "",
                 result_limit: int = Query(default=1000, le=10000)):
    return ok(data=V2RevenueService().transactions(advisor_id, month_id, group_id, result_limit))


@router.get("/evidence/reproduce")
def evidence_reproduce(advisor_id: str, product_group: str, from_month: str, to_month: str):
    return ok(data=V2RevenueService().product_revenue_change(
        advisor_id, product_group, from_month, to_month))


# ---------------------------------------------------------------- ops

@router.get("/ops/counts")
def ops_counts():
    return ok(data=V2RevenueService().ingestion_counts())


@router.get("/ops/advisor-summary")
def ops_advisor_summary(advisor_id: str):
    return ok(data=V2RevenueService().advisor_month_summary(advisor_id))


@router.get("/ops/reconciliation")
def ops_reconciliation(advisor_id: str, from_month: str, to_month: str):
    return ok(data=V2DriverService().reconciliation(advisor_id, from_month, to_month))

File: frontend/lib/api/v2.ts
/**
 * Typed fetchers for the V2 API (/api/v2/*). Every payload carries
 * served_by_tier: 1 = TigerGraph, 2 = local store — the context-bar tier pill
 * reads it so the UI never misrepresents where a number came from.
 */
import { apiClient } from "@/lib/api/client";

export type Provenance = "REAL" | "DERIVED" | "ASSUMED" | "DUMMY";

export interface Advisor {
  advisor_sid: string;
  advisor_name: string;
  rep_code: string;
  branch_cd: string;
  data_source: Provenance;
}

export interface MonthRow {
  month_id: string;
  year: number;
  month_no: number;
  month_name: string;
  quarter: number;
  start_dt: string;
  end_dt: string;
  calendar_days: number;
  billable_days: number;
  prior_month_id: string;
  index_return: number;
  is_current: boolean;
  data_source: Provenance;
}

export interface HierarchyNode {
  parent_id: string;
  data_source: Provenance;
  class_id?: string;
  class_name?: string;
  line_id?: string;
  line_name?: string;
  group_id?: string;
  group_name?: string;
  product_id?: string;
  product_name?: string;
  display_order?: number;
}

export interface ProductHierarchy {
  classes: HierarchyNode[];
  lines: HierarchyNode[];
  groups: HierarchyNode[];
  products: HierarchyNode[];
  served_by_tier: number;
}

export interface DriverCause {
  cause_id: string;
  cause_name: string;
  cause_description: string;
  default_data_source: Provenance;
  display_order: number;
}

export interface MonthlyRevenueRow {
  mpr_id: string;
  advisor_sid: string;
  month_id: string;
  group_id: string;
  line_id: string;
  class_id: string;
  revenue: number;
  txn_count: number;
  account_count: number;
  avg_rate_bps: number;
  recurring_amt: number;
  one_time_amt: number;
  /** R1-6 credited-definition breakdown carried on every mpr row. */
  total_revenue?: number;
  non_credited_amt?: number;
  excluded_amt?: number;
  late_excluded_amt?: number;
  data_source: Provenance;
}

export interface RevenueChangeRow {
  change_id: string;
  advisor_sid: string;
  from_month_id: string;
  to_month_id: string;
  group_id: string;
  from_revenue: number;
  to_revenue: number;
  change_amt: number;
  change_pct: number;
  direction: "UP" | "DOWN" | "FLAT";
  data_source: Provenance;
}

export interface DriverRow {
  driver_id: string;
  change_id: string;
  cause_id: string;
  group_id: string;
  contribution_amt: number;
  contribution_pct: number;
  direction: "UP" | "DOWN";
  rank: number;
  inputs_json: string;
  data_source: Provenance;
}

export interface CommentaryBullet {
  driver_id: string;
  direction: "UP" | "DOWN";
  title: string;
  text: string;
  cause_id: string;
  group_id: string;
  data_source: Provenance;
}

export interface CommentaryRow {
  commentary_id: string;
  version_id: string;
  advisor_sid: string;
  from_month_id: string;
  to_month_id: string;
  headline: string;
  narrative_text: string;
  bullets_json: string;
  status: "PUBLISHED" | "BLOCKED";
  blocked_reason: string;
  data_source: Provenance;
}

export interface CommentaryVersion {
  version_id: string;
  version_no: number;
  generated_at: string;
  model: string;
  prompt_version: string;
  data_snapshot_dt: string;
  status: "DRAFT" | "PUBLISHED" | "SUPERSEDED";
  advisor_count: number;
  transition_count: number;
  blocked_count: number;
  notes: string;
}

/** Independent LLM-judge review of one commentary (FIX_SPEC R5). Advisory —
 * deterministic guardrails remain the blocking gate. */
export interface CommentaryEvaluation {
  evaluation_id: string;
  commentary_id: string;
  version_id: string;
  judge_model: string;
  faithfulness_score: number; // 0-1
  hallucination_flag: boolean;
  completeness_score: number;
  clarity_score: number;
  verdict: "PASS" | "REVIEW" | "FAIL";
  reasoning: string;
  evaluated_at: string;
  data_source: Provenance;
}

export interface EvidenceRecord {
  evidence_id: string;
  driver_id: string;
  finding_text: string;
  calc_json: string;
  source_records_json: string;
  lineage_json: string;
  checks_json: string;
  gsql_query_name: string;
  gsql_params_json: string;
  gsql_result_json: string;
  source_sql: string;
  source_table: string;
  source_row_count: number;
  data_source: Provenance;
}

export interface TransactionRow {
  txn_id: string;
  trade_ref_no: string;
  trade_dt: string;
  product_id: string;
  product_name: string;
  group_id: string;
  account_no: string;
  credited_amt: number;
  split_pct: number;
  file_key: string;
  rev_nature: string;
  month_id: string;
  /** R1/R3 — credited-eligibility classification of the row. */
  eligibility_bucket: "CREDITED" | "NON_CREDITED" | "EXCLUDED" | "LATE" | "OUT_OF_GRID";
  reason_cd: string; // "__NONE__" when the row carries no reason code
  grid_type: string;
  data_source: Provenance;
}

export interface MonthlyTotals {
  revenue_by_month: Record<string, number>;
  recurring_by_month: Record<string, number>;
  non_recurring_by_month: Record<string, number>;
  txn_count_by_month: Record<string, number>;
  served_by_tier: number;
}

/** R6 Y — stored anomaly row (detection is batch; the screen only retrieves). */
export interface AnomalyRow {
  anomaly_id: string;
  advisor_sid: string;
  from_month_id: string;
  to_month_id: string;
  rule_id: string;
  severity: "HIGH" | "MEDIUM" | "LOW" | "INFO";
  title: string;
  detail_text: string;
  metrics_json: string;
  threshold_json: string;
  impact_amt: number;
  group_id: string;
  scan_id: string;
  detected_at: string;
  data_source: Provenance;
}

export interface AnomalyScan {
  scan_id: string;
  started_at: string;
  advisors_reviewed: number;
  transitions_reviewed: number;
  flagged_count: number;
  thresholds_json: string;
  status: string;
}

export interface AnomaliesResponse {
  scan_id_used: string;
  scan: Partial<AnomalyScan>;
  anomalies: AnomalyRow[];
  thresholds_in_force: Record<string, number>;
  served_by_tier: number;
}

export const v2Api = {
  advisors: () => apiClient.get<{ advisors: Advisor[]; served_by_tier: number }>("/api/v2/reference/advisors"),
  months: () => apiClient.get<{ months: MonthRow[]; served_by_tier: number }>("/api/v2/reference/months"),
  productHierarchy: () => apiClient.get<ProductHierarchy>("/api/v2/reference/product-hierarchy"),
  driverCauses: () => apiClient.get<{ causes: DriverCause[]; served_by_tier: number }>("/api/v2/reference/driver-causes"),

  trendsRevenue: (advisorId: string, fromMonth: string, toMonth: string) =>
    apiClient.get<{ monthly_revenue: MonthlyRevenueRow[]; served_by_tier: number }>(
      `/api/v2/trends/revenue?advisor_id=${advisorId}&from_month=${fromMonth}&to_month=${toMonth}`),
  trendsChanges: (advisorId: string, fromMonth: string, toMonth: string) =>
    apiClient.get<{ changes: RevenueChangeRow[]; served_by_tier: number }>(
      `/api/v2/trends/changes?advisor_id=${advisorId}&from_month=${fromMonth}&to_month=${toMonth}`),

  insightsChart: (advisorId: string, fromMonth: string, toMonth: string) =>
    apiClient.get<MonthlyTotals>(
      `/api/v2/insights/chart?advisor_id=${advisorId}&from_month=${fromMonth}&to_month=${toMonth}`),
  insightsDrivers: (advisorId: string, fromMonth: string, toMonth: string, limit = 100) =>
    apiClient.get<{ drivers: DriverRow[]; served_by_tier: number }>(
      `/api/v2/insights/drivers?advisor_id=${advisorId}&from_month=${fromMonth}&to_month=${toMonth}&result_limit=${limit}`),
  commentary: (advisorId: string, versionId = "") =>
    apiClient.get<{ commentaries: CommentaryRow[]; resolved_version: string; served_by_tier: number; empty_state: string | null }>(
      `/api/v2/insights/commentary?advisor_id=${advisorId}&version_id=${versionId}`),
  versions: () =>
    apiClient.get<{ versions: CommentaryVersion[]; served_by_tier: number }>("/api/v2/insights/versions"),
  evaluations: (versionId: string) =>
    apiClient.get<{ evaluations: CommentaryEvaluation[]; served_by_tier: number }>(
      `/api/v2/insights/evaluations?version_id=${encodeURIComponent(versionId)}`),
  generate: (notes = "") => apiClient.post<Record<string, unknown>>(`/api/v2/insights/generate?notes=${encodeURIComponent(notes)}`),

  evidence: (driverId: string, versionId = "") =>
    apiClient.get<{ evidence: EvidenceRecord[]; served_by_tier: number }>(
      `/api/v2/evidence?driver_id=${encodeURIComponent(driverId)}&version_id=${versionId}`),

  transactions: (advisorId: string, monthId: string, groupId = "", limit = 1000) =>
    apiClient.get<{ transactions: TransactionRow[]; row_count: number; credited_total: number; served_by_tier: number }>(
      `/api/v2/transactions?advisor_id=${advisorId}&month_id=${monthId}&group_id=${encodeURIComponent(groupId)}&result_limit=${limit}`),

  opsCounts: () =>
    apiClient.get<{ counts: Record<string, number>; source_mix: Record<string, Record<string, number>>; served_by_tier: number }>(
      "/api/v2/ops/counts"),
  reconciliation: (advisorId: string, fromMonth: string, toMonth: string) =>
    apiClient.get<{ all_reconcile: boolean; transitions: Record<string, { total_change: number; attributed: number; discrepancy: number; reconciles: boolean }>; served_by_tier: number }>(
      `/api/v2/ops/reconciliation?advisor_id=${advisorId}&from_month=${fromMonth}&to_month=${toMonth}`),

  anomalies: (advisorId = "", scanId = "", severity = "", limit = 500) =>
    apiClient.get<AnomaliesResponse>(
      `/api/v2/anomalies?advisor_id=${encodeURIComponent(advisorId)}&scan_id=${encodeURIComponent(scanId)}&severity=${encodeURIComponent(severity)}&result_limit=${limit}`),
  anomalyScans: () =>
    apiClient.get<{ scans: AnomalyScan[]; served_by_tier: number }>("/api/v2/anomalies/scans"),
  anomalyScan: (notes = "") =>
    apiClient.post<Record<string, unknown>>(`/api/v2/anomalies/scan?notes=${encodeURIComponent(notes)}`),

  adapterStatus: () =>
    apiClient.get<{ modes: { graph_client_mode: string; llm_client_mode: string; data_set: string; commentary_mode: string } }>(
      "/adapters/status"),
};
